EventListener/TagSubscriber.php

```diff
@@ -73,24 +73,21 @@ public function onKernelResponse(FilterResponseEvent $event)
 
         $configuredTags = $this->matchRule($request, $response);
         if ($configuredTags) {
-            foreach ($configuredTags['tags'] as $tag) {
-                $tags[] = $tag;
-            }
+            $tags = array_merge($tags, $configuredTags['tags']);
             foreach ($configuredTags['expressions'] as $expression) {
                 $tags[] = $this->evaluateTag($expression, $request);
             }
         }
 
-        if (!count($tags)) {
-            return;
-        }
-
         if ($request->isMethodSafe()) {
-            // For safe requests (GET and HEAD), set cache tags on response
-            $this->cacheManager->tagResponse($response, $tags);
-        } else {
+            $this->tagHandler->addTags($tags);
+            if (HttpKernelInterface::MASTER_REQUEST === $event->getRequestType()) {
+                // For safe requests (GET and HEAD), set cache tags on response
+                $this->tagHandler->tagResponse($response);
+            }
+        } elseif (count($tags)) {
             // For non-safe methods, invalidate the tags
-            $this->cacheManager->invalidateTags($tags);
+            $this->tagHandler->invalidateTags($tags);
```

Comment:
Yes, this is unfortunate. It’s due to the `TagHandler` having two (albeit closely related) responsibilities instead of one: both tagging safe responses and invalidating unsafe ones. I don’t see how we can clear this up without splitting the class into two, which may be overkill. 

As for resetting: is our current `TagHandler` suitable for tagging multiple responses? Magically resetting between tag actions may also be confusing if a user wants to tag two responses identically.